### PR TITLE
Controller implementation

### PR DESCRIPTION
**Pull request by Ranray**

### CODE REVIEW

File: Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java [hunk 1]
```diff
@@ -62,29 +63,19 @@
...
-     * Method that checks if the given key is already present in pressedKeys
-     * list.
-     *
-     * @param code The keycode that needs to be checked.
-     * @return boolean True if already pressed, false if not in pressedKeys
-     * list.
-     */
-    public static boolean isPressed(KeyCode code) {
-        return PRESSED_KEYS.contains(code);
+    public Keyboard() {
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java]: MAJOR Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1186)

File: Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java [hunk 2]
```diff
@@ -17,6 +17,7 @@
 
     // Stores the keys that are being pressed at any time.
     private static final List<KeyCode> PRESSED_KEYS;
+    private static final InputController controller = InputController.getInstance();
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java]: MAJOR Remove this unused "controller" private field. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1068)

File: Soccar [UI]/src/nl/soccar/ui/input/InputController.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    private void poll() {
+        if (gamePad == null) {
+            pollKeyboard();
+        } else if (gamePad.isConnected()) {
+            setThrottleAction(gamePad.getThrottleAction());
+            setHandbrakeAction(gamePad.getHandbrakeAction());
+            setSteerAction(gamePad.getSteerAction());
+        } else {
```
Comment (ghost) [inline Soccar [UI]/src/nl/soccar/ui/input/InputController.java:58]: MAJOR This branch's code block is the same as the block for the branch on line 52. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1871)

File: Soccar [UI]/src/nl/soccar/ui/input/GamePad.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+    private ThrottleAction throttleAction;
+    private HandbrakeAction handbrakeAction;
+    private SteerAction steerAction;
+
+    private boolean POLL = false;
+    private Timer timer = new Timer(true);
+    private int xAxisPercentage;
+
+    public GamePad() {
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/input/GamePad.java]: MINOR Document this public constructor. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [UI]/src/nl/soccar/ui/input/GamePad.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        throttleAction = ThrottleAction.IDLE;
+        handbrakeAction = HandbrakeAction.INACTIVE;
+        steerAction = SteerAction.NONE;
+    }
+
+    public GamePad(Controller controller) {
```
Comment (ghost) [inline Soccar [UI]/src/nl/soccar/ui/input/GamePad.java:39]: MINOR Document this public constructor. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [UI]/src/nl/soccar/ui/input/GamePad.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        if (componentIdentifier == Component.Identifier.Button._0 && isPressed) {
+            throttleAction = ThrottleAction.BOOST;
+        } else if (componentIdentifier == Component.Identifier.Button._0 && !isPressed && throttleAction != ThrottleAction.ACCELERATE && throttleAction != ThrottleAction.REVERSE) {
+            throttleAction = ThrottleAction.IDLE;
+        }
+
+    }
+
+    private void processBrakeAction(Component.Identifier componentIdentifier, boolean isPressed) {
+        System.out.println(componentIdentifier.toString());
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/input/GamePad.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS106)

File: Soccar [UI]/src/nl/soccar/ui/input/GamePad.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    private static final int ACTION_REVERSE = 55;
+    private static final int POLL_TIME = 50;
+
+    private Controller controller;
+
+    private ThrottleAction throttleAction;
+    private HandbrakeAction handbrakeAction;
+    private SteerAction steerAction;
+
+    private boolean POLL = false;
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/input/GamePad.java]: MINOR Rename this field "POLL" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS00116)

File: Soccar [UI]/src/nl/soccar/ui/input/GamePad.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+            steerAction = SteerAction.NONE;
+        }
+    }
+
+    private void processButtonActions(Component c, Component.Identifier componentIdentifier) {
+        if (!componentIdentifier.getName().matches("^[0-9]*$")) {
+            return;
+        }
+
+        boolean isPressed = c.getPollData() != 0.0F;
```
Comment (TehMessiah) [inline Soccar [UI]/src/nl/soccar/ui/input/GamePad.java]: CRITICAL Equality tests should not be made with floating point values. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1244)

**Comment by TehMessiah** [comment]
SonarQube analysis reported 11 issues
* CRITICAL 1 critical
* MAJOR 4 major
* MINOR 6 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [CarUiFx.java#L22](https://github.com/PTS3-S34A/User-Interface/blob/95749c727928f237d51a54ece6b8935ef33ffbc8/Soccar [UI]/src/nl/soccar/ui/fx/drawable/CarUiFx.java#L22): Remove this unused import 'nl.soccar.ui.input.Keyboard'. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: Soccar [UI]/src/nl/soccar/ui/input/GamePad.java (hunk 7 above)
Comment (Laugslander) [inline Soccar [UI]/src/nl/soccar/ui/input/GamePad.java]: Waarom is POLL met hoofdletters?

File: Soccar [UI]/src/nl/soccar/ui/input/GamePad.java [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        if (!POLL) {
+            return;
+        }
+        if (!controller.poll()) {
+            POLL = false;
+            return;
+        }
+
+        Component[] components = controller.getComponents();
+        for (Component c : components) {
```
Comment (Laugslander) [inline Soccar [UI]/src/nl/soccar/ui/input/GamePad.java]: Kan in principe ook `for (Component c : controller.getComponents())` zijn omdat components maar één keer wordt gebruikt.

File: Soccar [UI]/src/nl/soccar/ui/input/GamePad.java [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        processBrakeAction(componentIdentifier, isPressed);
+    }
+
+    private void processBoostAction(Component.Identifier componentIdentifier, boolean isPressed) {
+        if (componentIdentifier == Component.Identifier.Button._0 && isPressed) {
+            throttleAction = ThrottleAction.BOOST;
+        } else if (componentIdentifier == Component.Identifier.Button._0 && !isPressed && throttleAction != ThrottleAction.ACCELERATE && throttleAction != ThrottleAction.REVERSE) {
+            throttleAction = ThrottleAction.IDLE;
+        }
+
```
Comment (Laugslander) [inline Soccar [UI]/src/nl/soccar/ui/input/GamePad.java]: Lege regel kan weg.

File: Soccar [UI]/src/nl/soccar/ui/input/GamePad.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    /**
+     * Given value of axis in percentage. Percentages increases from left/top to
+     * right/bottom. If idle (in center) returns 50, if joystick axis is pushed
+     * to the left/top edge returns 0 and if it's pushed to the right/bottom
+     * returns 100.
+     *
+     * @return value of axis in percentage.
+     */
+    private int getAxisValueInPercentage(float axisValue) {
+        return (int) (((2 - (1 - axisValue)) * 100) / 2);
```
Comment (Laugslander) [inline Soccar [UI]/src/nl/soccar/ui/input/GamePad.java]: Twee haakjes kunnen weg:
`return (int) ((2 - (1 - axisValue)) * 100 / 2);`

File: Soccar [UI]/src/nl/soccar/ui/input/InputController.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        start();
+    }
+
+    private void start() {
+        timer.scheduleAtFixedRate(new TimerTask() {
+            @Override
+            public void run() {
+                poll();
+            }
+        }, 0, 50);
```
Comment (Laugslander) [inline Soccar [UI]/src/nl/soccar/ui/input/InputController.java]: Magic value 50 toevoegen als constante.

File: Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java [hunk 13]
```diff
@@ -93,18 +84,18 @@ public static void setKeyPressed(KeyCode code) {
      *
      * @param code The keycode that needs to be removed of the pressedKeys list.
      */
-    public static void setKeyReleased(KeyCode code) {
-        PRESSED_KEYS.remove(code);
+    public void setKeyReleased(KeyCode code) {
```
Comment (Laugslander) [inline Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java:85]: Synchronisatie niet nodig.

File: Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java [hunk 14]
```diff
@@ -62,29 +63,19 @@
...
     /**
      * Method that adds the given KeyCode to the pressedKeys list.
      *
      * @param code The keycode that needs to be added to the pressedKeys list.
      */
-    public static void setKeyPressed(KeyCode code) {
-        if (!PRESSED_KEYS.contains(code)) {
-            PRESSED_KEYS.add(0, code); // Prepend to the list, so the last key pressed gets first priority
+    public void setKeyPressed(KeyCode code) {
+        synchronized (PRESSED_KEYS) {
```
Comment (Laugslander) [inline Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java]: Synchronisatie niet nodig.

File: Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java [hunk 15]
```diff
@@ -143,16 +133,15 @@ public static SteerAction getSteerAction() {
...
         }
-
         return SteerAction.NONE;
     }
 
     /**
      * Returns the correct HandbrakeAction based on the current pressed keys
      *
-     * @return SteerAction
+     * @return HandbtakeAction
```
Comment (Laugslander) [inline Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java]: Handb**r**akeAction

File: Soccar [UI]/src/nl/soccar/ui/input/InputController.java [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+    private GamePad gamePad;
+    private final Timer timer = new Timer(true);
+    
+    private InputController() {
+        throttleAction = ThrottleAction.IDLE;
+        handbrakeAction = HandbrakeAction.INACTIVE;
+        steerAction = SteerAction.NONE;
+    }
+
+    public void initializeInput(Keyboard keyboard) {
```
Comment (ghost) [inline Soccar [UI]/src/nl/soccar/ui/input/InputController.java:37]: MINOR Document this public method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [UI]/src/nl/soccar/ui/input/GamePad.java [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+    private ThrottleAction throttleAction;
+    private HandbrakeAction handbrakeAction;
+    private SteerAction steerAction;
+
+    private boolean isPolling = false;
+    private Timer timer = new Timer(true);
+    private int xAxisPercentage;
+
+    public GamePad() {
```
Comment (ghost) [inline Soccar [UI]/src/nl/soccar/ui/input/GamePad.java:33]: MINOR Document this public constructor. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

**Comment by ghost** [comment]
SonarQube analysis reported 4 issues
* MAJOR 1 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

File: Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java [hunk 18]
```diff
@@ -143,16 +133,15 @@ public static SteerAction getSteerAction() {
...
      */
-    public static HandbrakeAction getHandbrakeAction() {
+    HandbrakeAction getHandbrakAction() {
```
Comment (Laugslander) [inline Soccar [UI]/src/nl/soccar/ui/input/Keyboard.java:141]: getHandbrak**e**Action

**Review by Laugslander — CHANGES_REQUESTED** [review]

**Review by Laugslander — APPROVED** [review]

**Pull request merged**